File: src/main/java/no/fint/provider/springer/model/ArkivKodeverkRepository.java
package no.fint.provider.springer.model;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.event.model.ResponseStatus;
import no.fint.event.model.Status;
import no.fint.model.arkiv.kodeverk.KodeverkActions;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.arkiv.kodeverk.*;
import no.fint.provider.springer.storage.SpringerRepository;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Repository;

import jakarta.annotation.PostConstruct;
import java.util.EnumMap;
import java.util.Set;
import java.util.stream.Collectors;

@Slf4j
@Repository
public class ArkivKodeverkRepository extends SpringerRepository {

    private final EnumMap<KodeverkActions, Class<? extends FintLinks>> actions = new EnumMap<>(KodeverkActions.class);

    @PostConstruct
    public void init() {
        actions.put(KodeverkActions.GET_ALL_SAKSSTATUS, SaksstatusResource.class);
        actions.put(KodeverkActions.GET_ALL_SKJERMINGSHJEMMEL, SkjermingshjemmelResource.class);
        actions.put(KodeverkActions.GET_ALL_TILGANGSRESTRIKSJON, TilgangsrestriksjonResource.class);
        actions.put(KodeverkActions.GET_ALL_KLASSIFIKASJONSTYPE, KlassifikasjonstypeResource.class);
        actions.put(KodeverkActions.GET_ALL_DOKUMENTSTATUS, DokumentStatusResource.class);
        actions.put(KodeverkActions.GET_ALL_DOKUMENTTYPE, DokumentTypeResource.class);
        actions.put(KodeverkActions.GET_ALL_JOURNALSTATUS, JournalStatusResource.class);
        actions.put(KodeverkActions.GET_ALL_VARIANTFORMAT, VariantformatResource.class);
        actions.put(KodeverkActions.GET_ALL_ROLLE, RolleResource.class);
    }

    @Override
    public void accept(Event<FintLinks> response) {
        if (!StringUtils.equals(response.getSource(), "arkiv-kodeverk")) {
            log.info("Skipping {} from {}", response.getAction(), response.getSource());
            return;
        }

        KodeverkActions action = KodeverkActions.valueOf(response.getAction());
        if (actions.containsKey(action)) {
            query(actions.get(action), response);
        } else {
            response.setStatus(Status.ADAPTER_REJECTED);
            response.setResponseStatus(ResponseStatus.REJECTED);
            response.setStatusCode("INVALID_ACTION");
            response.setMessage("Invalid action: " + action);
        }
    }

    @Override
    public Set<String> actions() {
        return actions.keySet()
                .stream()
                .map(Enum::name)
                .collect(Collectors.toSet());
    }

}

File: src/main/kotlin/no/fint/provider/springer/model/ArkivKodeverkRepository.kt
package no.fint.provider.springer.model

import no.fint.event.model.Event
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.storage.SpringerRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.arkiv.kodeverk.KodeverkActions
import no.novari.fint.model.resource.arkiv.kodeverk.*
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository
import java.util.stream.Collectors
import java.util.stream.Stream

@Repository
class ArkivKodeverkRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper
) : SpringerRepository(mongoTemplate, wrapper) {

    override fun accept(response: Event<FintLinks>) {
        when {
            KodeverkActions.getActions().contains(response.action) -> when (KodeverkActions.valueOf(response.action)) {
                KodeverkActions.GET_ALL_SAKSSTATUS -> query(SaksstatusResource::class.java, response)
                KodeverkActions.GET_ALL_SKJERMINGSHJEMMEL -> query(SkjermingshjemmelResource::class.java, response)
                KodeverkActions.GET_ALL_TILGANGSRESTRIKSJON -> query(TilgangsrestriksjonResource::class.java, response)
                KodeverkActions.GET_ALL_KLASSIFIKASJONSTYPE -> query(KlassifikasjonstypeResource::class.java, response)
                KodeverkActions.GET_ALL_DOKUMENTSTATUS -> query(DokumentStatusResource::class.java, response)
                KodeverkActions.GET_ALL_DOKUMENTTYPE -> query(DokumentTypeResource::class.java, response)
                KodeverkActions.GET_ALL_JOURNALSTATUS -> query(JournalStatusResource::class.java, response)
                KodeverkActions.GET_ALL_VARIANTFORMAT -> query(VariantformatResource::class.java, response)
                KodeverkActions.GET_ALL_ROLLE -> query(RolleResource::class.java, response)
                else -> invalid(response)
            }
            else -> invalid(response)
        }
    }

    private fun invalid(response: Event<FintLinks>) {
        response.status = Status.ADAPTER_REJECTED
        response.responseStatus = ResponseStatus.REJECTED
        response.statusCode = "INVALID_ACTION"
        response.message = "Invalid action"
    }

    override fun actions(): Set<String> =
        Stream.of(*KodeverkActions.values())
            .map { it.name }
            .filter { it.startsWith("GET_ALL_") }
            .collect(Collectors.toSet())
}
